### DISCUSSION TITLE
large root.tsx file

### DISCUSSION THREAD

**Discussion by zhex900** [epicweb-dev/epic-stack#417] [Q&A]
Hi,

I want to learn the design principles of this template. 

`root.tsx` is over 300 lines. Is there a reason for keeping all the functions in one file? Why don't you split each function in separate a file and import it?

**Comment by kentcdodds** [comment]
This would be a great question for [the Call Kent Podcast](https://kcd.im/calls). Would you be willing to ask there?

**Comment by chessydk** [comment]
I must admit I have the same thoughs on root.tsx - not only the length, but separation of concerns and a bit around the functionality, so I'll put in some notes here:

- As the theme information just lives in a cookie, I don't see any reason for posting a change of mode to the server, thereby invalidating data so all loaders are re-executed without reason. Couldn't the theme cookie just be updated instantly on client, making optimistic dataLoading via useFetchers unnecessary ? 

- I think it would be great to pull all theme handing from root.tsx (except useTheme and setting the class of course). 

One possibility would be to have a pathless theme route that wraps the app. The userPrefs.theme property of the root loader would of course be moved to the new loader. But it is just annoying to have this composition of functionality exposed in the routes I think, so don't like it.

Another solution would be to use a "Fullstack component" like you suggest in your fine article: https://www.epicweb.dev/full-stack-components. Only thing is, we need to make an immediate submit on the fetcher on the server, but not on the client - so there would be some logic making sure it only runs on server.

Just my 5 pennies - thx. for your awesome inspiration!

**Reply by kentcdodds** [reply]
You could definitely just set the theme in the client. One edge case you'd run into is if any loader depends on the theme then that data will be stale when the theme is changed. Because users change the theme so infrequently, I don't mind the revalidation in favor of correctness over the optimization.

I've had the theme in a separate file before. I don't feel strongly about it. A PR would be welcome if anyone feels strongly about it.
